src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -146,25 +161,29 @@ class HistoryWriterImpl private(file: Option[File], val synchronizationToken: Re
 
   override def close(): Unit = db.close()
 
+  override def featureActivationHeight(feature: Short): Option[Int] = read { implicit lock =>
+    functionalitySettings.preActivatedFeatures.get(feature)
+      .orElse(Option(featuresState().get(feature)).map(h => h + ActivationWindowSize))
+  }
+
+  override def featureStatus(feature: Short): BlockchainFeatureStatus = read { implicit lock =>
+    Option(featuresState().get(feature))
+      .map(h => if (h <= height - ActivationWindowSize)
+        BlockchainFeatureStatus.Activated else
+        BlockchainFeatureStatus.Accepted)
+      .getOrElse(BlockchainFeatureStatus.Undefined)
+  }
+
+  override def activationWindowOpeningFromHeight(height: Int): Int = {
```

Comment:
this is pure function, never really overridden, should be placed in companion object